src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,426 @@
+                {
+                    nodeFound = true;
+                }
+
+                // No special action is performed at this time if `;` is typed inside a string, including
+                // interpolated strings.  
+                if (IsInAString(currentNode, isCaretAtEndOfLine))
+                {
+                    return false;
+                }
+
+                // We reached the root without finding a statement
+                if (currentNode.Parent == null)
+                {
+                    return false;
+                }
+
+                currentNode = currentNode.Parent;
+            }
+
+            // if we never found a statement, or a node of interest, or the statement kind is not a candidate for completion, return
+            if (currentNode == null || !nodeFound || !StatementIsACandidate(currentNode, caret))
+            {
+                return false;
+            }
+
+            Debug.Assert(currentNode != null);
+            return true;
+        }
+
+        private static bool IsStatementOrFieldDeclaration(SyntaxNode currentNode, ISyntaxFactsService syntaxFacts)
+            => syntaxFacts.IsStatement(currentNode) || currentNode.IsKind(SyntaxKind.FieldDeclaration);
+
+        private static bool IsInAString(SyntaxNode currentNode, bool isCaretAtEndOfLine)
+                     // If caret is at the end of the line, it is outside the string	
+                     => (currentNode.IsKind(SyntaxKind.InterpolatedStringExpression, SyntaxKind.StringLiteralExpression)
+                         && !isCaretAtEndOfLine);
+
+        private static bool StatementIsACandidate(SyntaxNode currentNode, int caretPosition)
+        {
+            // if the statement kind ends in a semicolon, return true
+            switch (currentNode.Kind())
+            {
+                case SyntaxKind.DoStatement:
+                case SyntaxKind.ExpressionStatement:
+                case SyntaxKind.GotoCaseStatement:
+                case SyntaxKind.LocalDeclarationStatement:
+                case SyntaxKind.ReturnStatement:
+                case SyntaxKind.ThrowStatement:
+                case SyntaxKind.FieldDeclaration:
+                    return true;
+                case SyntaxKind.ForStatement:
+                    var forStatementSyntax = (ForStatementSyntax)currentNode;
+                    if (CaretIsInForStatementCondition(caretPosition, forStatementSyntax) ||
+                        CaretIsInForStatementDeclaration(caretPosition,forStatementSyntax))
+                    {
+                        return true;
+                    }
+                    else
+                    {
+                        return false;
+                    }
+                default:
+                    return false;
+            }
+        }
+
+        private static bool SemicolonIsMissing(SyntaxNode currentNode)
+        {
+            switch (currentNode.Kind())
```

Comment:
this is the one. something we can make a generic helper like GetSemicolon?

Reply:
Yes, this is addressed in follow-up issue https://github.com/dotnet/roslyn/issues/32163